### PR TITLE
OpenApi + Ktor + Resources

### PR DESCRIPTION
**Pull request by serras**
This PR introduces a new module which integrates `tegral-openapi-ktor` with [type-safe routing](https://ktor.io/docs/type-safe-routing.html). With this package, the OpenAPI descriptions can be part of the resource definition:

```kotlin
@Resource("/hello/{name}")
class Hello(val name: String) {
    companion object: OpenApiDescription {
        override val openApi = { 
            description = "Returns a greeting"
        }
    }
}
```

There's a bit of ceremony in making the `companion object` implement an interface, but I couldn't find a better way to attach a piece of static information to the class, while at the same time giving access to the whole `OperationDsl`.

### CODE REVIEW

**Comment by utybo** [comment]
Hi! Thank you very much for your PR! This is a great idea. I have a few ideas:

Here's a possible alternative syntax to get less ceremony:

```kotlin
// In tegral-openapi-ktor-resources
fun describeResource(operationDsl: OperationDsl.() -> Unit): OpenApiDescription {
    return object : OpenApiDescription {
        override val openApi = operationDsl
    }
}

// In the "consumer" code
@Resource("/hello/{name}")
class Hello(val name: String) {
    companion object : OpenApiDescription by describeResource({
        summary = "Hello there"
    })
}
```

It saves 2 lines and an indentation, but I think it mostly helps for readability. The syntax is admittedly a bit weird (you can't just put `describeResource {`, the compiler will think that the `{` represents the beginning of the object body instead of being the lambda parameter of the `describeResource` function).

I also have some suggestions around the naming of the new symbols. I mostly want to avoid confusion with the non-resources variants for describing endpoints:

- `OpenApiDescription` -> `ResourceDescription`
- `description()` -> `descriptionFromCompanionObject()`

What do you think?

I can take care of making these changes, adding tests, docs and changelog entries if that's okay for you. My apologies for the lack of instructions/guidance around contributions, I have to admit I didn't expect people to find this project and contribute to it 😅

Also, as a later addition to Tegral, I think we can infer a lot of information from the things contained in the resource object (e.g. automatically create query parameters, path parameters, body information, etc. from the properties in the class). I'll create an issue about this once this is merged.

Once again, thank you for your contribution!

**Comment by serras** [comment]
@utybo Thanks to you for your nice project! I can tell you we were using it already under its previous name, because it requires so little additional preparation to get an OpenAPI spec from the Ktor routes.

I agree with all the changes you propose. If you don’t mind, I would prefer to keep both ways to add a description, both by overriding `openApi` and by delegating to `describeResource`.

I can take care of this tonight, or you can do it, as you prefer.

**Comment by utybo** [comment]
Your comment made my day, thank you!

> I agree with all the changes you propose. If you don’t mind, I would prefer to keep both ways to add a description, both by overriding openApi and by delegating to describeResource.

Sure thing!

I'll take care of making the necessary changes 👍🏻

File: tegral-openapi/tegral-openapi-ktor-resources/src/main/kotlin/guru/zoroark/tegral/openapi/ktor/resources/OpenApiKtorResources.kt [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        override val openApi = description
+    }
+}
+
+/**
+ * Retrieves a [resource description][ResourceDescription] from the given class' companion object.
+ */
+inline fun <reified T : Any> descriptionFromCompanionObject(): OperationDsl.() -> Unit =
+    (T::class.companionObjectInstance as? ResourceDescription)?.openApi
+        ?: { }
```
Comment (utybo) [inline tegral-openapi/tegral-openapi-ktor-resources/src/main/kotlin/guru/zoroark/tegral/openapi/ktor/resources/OpenApiKtorResources.kt:81]: I'm not entirely sure of what should be done in this case. I'm torn between three options:

- Do nothing (which is what is done here)
- Log a warning via some logger
- Throw an exception

I'd prefer the last option, but I'm not sure if there's a valid case where `descriptionFromCompanionObject` is called on an "invalid"/badly-configured class...

(Two tests depend on the chosen behavior here, so I've `TODO()`-ed them)

**Review by utybo — COMMENTED** [review]
I've made the changes I mentioned, but there's still one point where I'm not sure of what to do, I've left a comment there

File: tegral-openapi/tegral-openapi-ktor-resources/src/main/kotlin/guru/zoroark/tegral/openapi/ktor/resources/OpenApiKtorResources.kt (hunk 1 above)
Comment (serras) [reply tegral-openapi/tegral-openapi-ktor-resources/src/main/kotlin/guru/zoroark/tegral/openapi/ktor/resources/OpenApiKtorResources.kt:81]: My use case for doing nothing was that when you integrate Tegral in your app, you start by replacing each `get`, `post`, ... by the corresponding version `getD`, `postD`, ... and then you slowly start adding information. This makes even more sense in a future in which the description already gets some information from the `@Resource` class itself.

Comment (utybo) [reply tegral-openapi/tegral-openapi-ktor-resources/src/main/kotlin/guru/zoroark/tegral/openapi/ktor/resources/OpenApiKtorResources.kt:81]: That's true, I didn't think of that. I'll update the tests then 👌🏻

**Comment by serras** [comment]
I've just pushed a new idea that came through my mind; if you prefer to not have it, feel free to remove it. The idea is to bring a bit more type-safety by allowing people to use a property reference instead of a name.

```kotlin
@Resource("/hello/{name}") class Hello(val name: String) {
  companion object : ResourceDescription {
    override val openApi = {
      pathParameter(Hello::name) {
        description = "The name"
      }
    }
  }
}
```

This may also be better as part of the DSL package itself, as it works better with Resources, but doesn't really depend on it.

**Comment by utybo** [comment]
I think they belong in the resources package, but that makes it annoyingly inconsistent with string-based parameters as the function can't be made infix (at least until we get [context receivers in Kotlin](https://github.com/Kotlin/KEEP/issues/259)).

I'd say these should be included at a later date (i.e. after this PR), and should probably wait for context receivers in Kotlin and the fancier "automagically infer OpenAPI information from resource classes" functionality. In the mean time, if people want these helpers, they can add their own extension functions (whose implementation is fairly trivial).

Also, I have been unable to use the syntax you've used in this PR for overriding the `openApi` property, I've always had to add the `: OperationDsl.() -> Unit` type after it, otherwise I'd get errors in IntelliJ. Is there something I'm missing here?

**Comment by serras** [comment]
Ok, I've now removed it.

It might well be the case that IntelliJ added the type for me, and I wasn't aware of that. In any case, I think this shows that `describeResource` works better :)

**Comment by utybo** [comment]
Everything looks good to me, I'll wait for the CI and preview website before merging

**Comment by serras** [comment]
One small question: how would you feel about changing the names of `getD`, `postD`, and so on to simply `get`, `post`...? So people would not import the `ktor.server.resources` package, but Tegral's one.

**Comment by utybo** [comment]
> One small question: how would you feel about changing the names of `getD`, `postD`, and so on to simply `get`, `post`...? So people would not import the `ktor.server.resources` package, but Tegral's one.

Apologies on the delay for the response. I'm a bit split on this one. On one hand it's a very interesting possibility for the future when "automatic documentation" happens (so that you can get documentation on resources "for free" by just changing your imports and nothing else), but it may end up being very confusing for end-users.

**Comment by utybo** [comment]
I'd say we should stay with the `xyzD` variants. I'm pretty sure the import clashing would be confusing for users (just like resources and non-resources of `get`/`post`/etc are already confusing). Making the difference obvious is probably for the better. I'll merge this PR today and try to get the 0.0.3 release done this weekend, once again apologies for the delay!

**Comment by utybo** [comment]
Merged, thank you for your contribution!

**Pull request merged**